Cannot use name of field converted with a `CompositeAttributeConverter` in converted map

<!-- ================================================================= -->
<!-- Please note that GitHub issues are only meant for bug reports/feature requests. -->
<!-- If you have questions on how to use the Neo4j OGM, please ask on -->
<!-- StackOverflow: http://stackoverflow.com/questions/tagged/neo4j-ogm instead of -->
<!-- creating an issue here. -->
<!-- ================================================================= -->

<!--- Provide a general summary of the issue in the Title above -->

In our project, we are trying to analyse source code in a graph structure. Each source code element is represented as a `Node`. Each `Node` has a `name` attribute, which is not a string, but a special `Name` class. This is primarily for reasons such as parsing of fully qualified names (e.g. `MyClass::test`). To serialise this special class, we have written a `CompositeAttributeConverter` that serialises the individual elements into a map of strings (`fullName`, `delimiter`, `localName`). However, we want also want to serialise a string representation into a property called `name` (the same as the original field name). While this works perfectly when serialising into the graph; it does not when we try to load a node again. 

## Expected Behavior

I would expect the OGM to recognise the field `name` as part of the converted set of attributes and that the `toEntityAttribute` is called.

## Current Behavior

Currently, OGM recognises `name` as a field property and tries to directly set the string value to the Java object, without consulting the composite attribute converter. If I remove the `name` key from the map, everything works.

```
Error mapping GraphModel
org.neo4j.ogm.exception.core.MappingException: Error mapping GraphModel
	at app//org.neo4j.ogm.context.GraphEntityMapper.mapContentOf(GraphEntityMapper.java:160)
	at app//org.neo4j.ogm.context.GraphEntityMapper.lambda$map$1(GraphEntityMapper.java:106)
	at app//org.neo4j.ogm.context.GraphEntityMapper.map(GraphEntityMapper.java:111)
	at app//org.neo4j.ogm.context.GraphRowModelMapper.map(GraphRowModelMapper.java:55)
	at app//org.neo4j.ogm.session.delegates.LoadByTypeDelegate.lambda$loadAll$0(LoadByTypeDelegate.java:108)
	at app//org.neo4j.ogm.session.Neo4jSession.doInTransaction(Neo4jSession.java:589)
	at app//org.neo4j.ogm.session.Neo4jSession.doInTransaction(Neo4jSession.java:563)
	at app//org.neo4j.ogm.session.delegates.LoadByTypeDelegate.loadAll(LoadByTypeDelegate.java:95)
	at app//org.neo4j.ogm.session.delegates.LoadByTypeDelegate.loadAll(LoadByTypeDelegate.java:115)
	at app//org.neo4j.ogm.session.Neo4jSession.loadAll(Neo4jSession.java:193)
	at app//de.fraunhofer.aisec.cpg_vis_neo4j.ApplicationTest.testPush(ApplicationTest.kt:71)
	at java.base@17.0.5/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base@17.0.5/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base@17.0.5/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base@17.0.5/java.lang.reflect.Method.invoke(Method.java:568)
	at app//org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:686)
	at app//org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at app//org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at app//org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at app//org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at app//org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at app//org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at app//org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at app//org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:212)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:208)
	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:137)
	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:71)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:135)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at app//org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at java.base@17.0.5/java.util.ArrayList.forEach(ArrayList.java:1511)
	at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at app//org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at java.base@17.0.5/java.util.ArrayList.forEach(ArrayList.java:1511)
	at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at app//org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at app//org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at app//org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:62)
	at java.base@17.0.5/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base@17.0.5/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base@17.0.5/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base@17.0.5/java.lang.reflect.Method.invoke(Method.java:568)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at jdk.proxy2/jdk.proxy2.$Proxy5.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:113)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:65)
	at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
Caused by: java.lang.IllegalArgumentException: Can not set de.fraunhofer.aisec.cpg.graph.Name field de.fraunhofer.aisec.cpg.graph.Node.name to java.lang.String
	at java.base/jdk.internal.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:167)
	at java.base/jdk.internal.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:171)
	at java.base/jdk.internal.reflect.UnsafeObjectFieldAccessorImpl.set(UnsafeObjectFieldAccessorImpl.java:81)
	at java.base/java.lang.reflect.Field.set(Field.java:799)
	at org.neo4j.ogm.metadata.FieldInfo.write(FieldInfo.java:421)
	at org.neo4j.ogm.metadata.FieldInfo.write(FieldInfo.java:446)
	at org.neo4j.ogm.context.GraphEntityMapper.writeProperty(GraphEntityMapper.java:293)
	at org.neo4j.ogm.context.GraphEntityMapper.setProperties(GraphEntityMapper.java:264)
	at org.neo4j.ogm.context.GraphEntityMapper.mapNodes(GraphEntityMapper.java:224)
	at org.neo4j.ogm.context.GraphEntityMapper.mapContentOf(GraphEntityMapper.java:150)
	... 95 more
```

The `writeProperty` function would even know that a composite converter is in place, however that information is not used (either in the function directly or above).

<img width="845" alt="Screenshot 2023-01-25 at 14 12 10" src="https://user-images.githubusercontent.com/12459061/214572173-3728591b-455f-4245-82a0-3488833a079d.png">

## Your Environment

- Java version: 17
- Neo4j version: 
- OGM Version: 4.0.1
- OGM Transport used: Bolt
- Neo4j Java Driver version: 4.0.1
- Link to your project: http://github.com/Fraunhofer-AISEC/cpg

--- a/core/src/main/java/org/neo4j/ogm/context/GraphEntityMapper.java
+++ b/core/src/main/java/org/neo4j/ogm/context/GraphEntityMapper.java
@@ -286,6 +286,8 @@ public class GraphEntityMapper {
 
         if (writer == null) {
             logger.debug("Unable to find property: {} on class: {} for writing", property.getKey(), classInfo.name());
+        } else if (writer.isComposite()) {
+            logger.info("Property {} is already handled by a CompositeAttributeConverter", property.getKey());
         } else {
             Object value = property.getValue();
             // merge iterable / arrays and co-erce to the correct attribute type

--- a/neo4j-ogm-tests/neo4j-ogm-integration-tests/src/test/java/org/neo4j/ogm/domain/gh932/EntityWithCompositeConverter.java
+++ b/neo4j-ogm-tests/neo4j-ogm-integration-tests/src/test/java/org/neo4j/ogm/domain/gh932/EntityWithCompositeConverter.java
@@ -1,0 +1,132 @@
+/*
+ * Copyright (c) 2002-2022 "Neo4j,"
+ * Neo4j Sweden AB [http://neo4j.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.neo4j.ogm.domain.gh932;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+import org.neo4j.ogm.annotation.GeneratedValue;
+import org.neo4j.ogm.annotation.Id;
+import org.neo4j.ogm.annotation.NodeEntity;
+import org.neo4j.ogm.annotation.typeconversion.Convert;
+import org.neo4j.ogm.typeconversion.CompositeAttributeConverter;
+
+/**
+ * @author Christian Banse
+ */
+@NodeEntity
+public class EntityWithCompositeConverter {
+
+    @Id
+    @GeneratedValue
+    private Long id;
+
+    @Convert(value = Converter.class)
+    private Name name;
+
+    public Name getName() {
+        return name;
+    }
+
+    public void setName(Name name) {
+        this.name = name;
+    }
+
+    public Long getId() {
+        return this.id;
+    }
+
+    @Override public boolean equals(Object o) {
+        if (this == o)
+            return true;
+        if (o == null || getClass() != o.getClass())
+            return false;
+        EntityWithCompositeConverter that = (EntityWithCompositeConverter) o;
+        return Objects.equals(id, that.id) && Objects.equals(name, that.name);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id, name);
+    }
+
+    public static class Name {
+        private String partialName1;
+        private String partialName2;
+
+        public String getPartialName1() {
+            return partialName1;
+        }
+
+        public String getPartialName2() {
+            return partialName2;
+        }
+
+        public void setPartialName1(String partialName1) {
+            this.partialName1 = partialName1;
+        }
+
+        public void setPartialName2(String partialName2) {
+            this.partialName2 = partialName2;
+        }
+
+        @Override
+        public String toString() {
+            return partialName1 + "." + partialName2;
+        }
+
+        @Override public boolean equals(Object o) {
+            if (this == o)
+                return true;
+            if (o == null || getClass() != o.getClass())
+                return false;
+            Name name = (Name) o;
+            return Objects.equals(partialName1, name.partialName1) && Objects.equals(partialName2, name.partialName2);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(partialName1, partialName2);
+        }
+    }
+
+    public static class Converter implements CompositeAttributeConverter<Name> {
+
+        @Override
+        public Map<String, ?> toGraphProperties(Name value) {
+            // We want to persist the partial properties of the name as well as a "toString" representation of the whole name as "name" (which "conflicts" with the "name" field on the class).
+            var map = new HashMap<String, String>();
+            map.put("name", value.toString());
+            map.put("partialName1", value.partialName1);
+            map.put("partialName2", value.partialName2);
+
+            return map;
+        }
+
+        @Override
+        public Name toEntityAttribute(Map<String, ?> value) {
+            var name = new Name();
+            name.partialName1 = String.valueOf(value.get("partialName1"));
+            name.partialName2 = String.valueOf(value.get("partialName2"));
+
+            return name;
+        }
+    }
+}

--- a/neo4j-ogm-tests/neo4j-ogm-integration-tests/src/test/java/org/neo4j/ogm/persistence/types/properties/PropertiesTest.java
+++ b/neo4j-ogm-tests/neo4j-ogm-integration-tests/src/test/java/org/neo4j/ogm/persistence/types/properties/PropertiesTest.java
@@ -29,15 +29,23 @@ import java.util.HashMap;
 import java.util.Iterator;
 import java.util.Map;
 
+import ch.qos.logback.classic.Level;
+import ch.qos.logback.classic.Logger;
+
 import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.neo4j.ogm.context.GraphEntityMapper;
+import org.neo4j.ogm.domain.gh932.EntityWithCompositeConverter;
 import org.neo4j.ogm.domain.properties.SomeNode;
 import org.neo4j.ogm.domain.properties.User;
 import org.neo4j.ogm.exception.core.MappingException;
 import org.neo4j.ogm.session.Session;
 import org.neo4j.ogm.session.SessionFactory;
+import org.neo4j.ogm.testutil.LoggerRule;
 import org.neo4j.ogm.testutil.TestContainersTestBase;
+import org.slf4j.LoggerFactory;
 
 /**
  * @author Frantisek Hartman
@@ -49,9 +57,12 @@ public class PropertiesTest extends TestContainersTestBase {
 
     private Session session;
 
+    @RegisterExtension
+    private final LoggerRule loggerRule = new LoggerRule();
+
     @BeforeAll
     public static void init() {
-        sessionFactory = new SessionFactory(getDriver(), User.class.getName(), SomeNode.class.getName());
+        sessionFactory = new SessionFactory(getDriver(), User.class.getName(), SomeNode.class.getName(), EntityWithCompositeConverter.class.getName());
     }
 
     @BeforeEach
@@ -459,5 +470,33 @@ public class PropertiesTest extends TestContainersTestBase {
 
         assertThat(user.getMyProperties())
             .containsOnlyKeys("prop1", "prop2");
+    }
+
+    // GH-932
+    @Test
+    void shouldResolveNameConflictInCompositeConverter() {
+
+        // ensure the right branch in the condition is hit by verifying the log message
+        Logger logger = (Logger) LoggerFactory.getLogger(GraphEntityMapper.class);
+        Level originalLevel = logger.getLevel();
+        logger.setLevel(Level.INFO);
+
+        try {
+            EntityWithCompositeConverter e = new EntityWithCompositeConverter();
+            EntityWithCompositeConverter.Name n = new EntityWithCompositeConverter.Name();
+            n.setPartialName1("some");
+            n.setPartialName2("entity");
+            e.setName(n);
+            session.save(e);
+            session.clear();
+
+            var loaded = session.load(EntityWithCompositeConverter.class, e.getId());
+            assertThat(loaded).isEqualTo(e);
+
+            assertThat(loggerRule.getFormattedMessages())
+                .containsSequence("Property name is already handled by a CompositeAttributeConverter");
+        } finally {
+            logger.setLevel(originalLevel);
+        }
     }
 }

